File: aoc/day1/day1.ml
let split_lines lines =
  let rec split acc_left acc_right = function
    | [] -> (List.rev acc_left, List.rev acc_right)
    | line :: rest ->
        let left, right = Scanf.sscanf line "%d %d" (fun l r -> (l, r)) in
        split (left :: acc_left) (right :: acc_right) rest
  in
  split [] [] lines

let puzzle1 left right =
  List.sort compare left |> fun sorted_left ->
  List.sort compare right |> fun sorted_right ->
  List.fold_left2 (fun acc x y -> acc + abs (x - y)) 0 sorted_left sorted_right

let build_hashmap list =
  let hashmap = Hashtbl.create (List.length list) in
  List.iter (fun n -> let count = try Hashtbl.find hashmap n with Not_found -> 0 in Hashtbl.replace hashmap n (count + 1))
  list; hashmap

let puzzle2 left hashmap_right =
  List.fold_left (fun acc n ->
    let count = try Hashtbl.find hashmap_right n with Not_found -> 0 in acc + (n * count))
    0 left

let solve filename =
  filename |> Read_file.read_file |> split_lines |> fun (left, right) ->
  let res1 = puzzle1 left right in
  Printf.printf "The Puzzle 1 solution is :%d\n" res1;

  right |> build_hashmap |> fun hashmap_right ->
  let res2 = puzzle2 left hashmap_right in
  Printf.printf "The Puzzle 2 solution is: %d\n" res2

File: aoc/day1/day1.py
f = open("input.txt", "r")
lines = f.readlines()
left, right = [], []

for l in lines:
    left_side, right_side = map(int, l.split())
    left.append(left_side)
    right.append(right_side)

left.sort()
right.sort()
res1 = 0

for x, y in zip(left, right):
    dist = abs(x - y)
    # print(dist)
    res1 += dist

print(f'The Puzzle 1 Solution is: {res1}')

# Part 2

# left = [3,4,2,1,3,3]
# right = [4,3,5,3,9,3]
hashmap_right = {}
for n in right:
    hashmap_right[n] = 1 + hashmap_right.get(n, 0)

res2 = 0
for n in left:
    res2 += n * hashmap_right.get(n, 0)

print(f'The Puzzle 2 Solution is: {res2}')